sql/catalyst/src/test/scala/org/apache/spark/sql/util/TimestampFormatterSuite.scala

```diff
@@ -291,4 +291,95 @@ class TimestampFormatterSuite extends SparkFunSuite with SQLHelper with Matchers
       }
     }
   }
+
+  test("parsing hour with various patterns") {
+    def createFormatter(pattern: String): TimestampFormatter = {
+      // Use `SIMPLE_DATE_FORMAT`, so that the legacy parser also fails with invalid value range.
+      TimestampFormatter(pattern, ZoneOffset.UTC, LegacyDateFormats.SIMPLE_DATE_FORMAT, false)
+    }
+
+    withClue("HH") {
+      val formatter = createFormatter("yyyy-MM-dd HH")
+
+      val micros1 = formatter.parse("2009-12-12 00")
+      assert(micros1 === TimeUnit.SECONDS.toMicros(
+        LocalDateTime.of(2009, 12, 12, 0, 0, 0).toEpochSecond(ZoneOffset.UTC)))
+
+      val micros2 = formatter.parse("2009-12-12 15")
+      assert(micros2 === TimeUnit.SECONDS.toMicros(
+        LocalDateTime.of(2009, 12, 12, 15, 0, 0).toEpochSecond(ZoneOffset.UTC)))
```

Comment:
I would propose to improve test coverage and test different years -100, 0001, 1582, 1900, 1970, 2037, 2100

Reply:
It's UT, which is whitebox testing. It's clear from the code that the time parsing is unrelated to the value of year